internal/errors/errors.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package errors defines the error handling used by Kpt codebase.
+package errors
+
+import (
+	"fmt"
+	"strings"
+)
+
+// Error is an implementation of the error interface used in the kpt
+// codebase.
+// It is based on the design in https://commandcenter.blogspot.com/2017/12/error-handling-in-upspin.html
+type Error struct {
+	// Path is the path name of the object involved in Kpt operation.
+	Path Path
+
+	// Op is the operation being performed, for ex. pkg.get, fn.render
+	Op Op
+
+	// Kind refers to classs of errors
+	Kind Kind
+
+	// Err refers to wrapped error (if any)
+	Err error
+}
+
+func (e *Error) Error() string {
+	b := new(strings.Builder)
+
+	if e.Op != "" {
+		pad(b, ": ")
+		b.WriteString(string(e.Op))
+	}
+
+	if e.Path != "" {
+		pad(b, ": ")
+		b.WriteString("pkg ")
+		b.WriteString(string(e.Path))
+	}
+
+	if e.Kind != 0 {
+		pad(b, ": ")
+		b.WriteString(e.Kind.String())
+	}
+
+	if e.Err != nil {
+		if wrappedErr, ok := e.Err.(*Error); ok {
+			if !wrappedErr.Zero() {
+				pad(b, ":\n\t")
+				b.WriteString(wrappedErr.Error())
+			}
+		} else {
+			pad(b, ": ")
+			b.WriteString(e.Err.Error())
+		}
+	}
+	if b.Len() == 0 {
+		return "no error"
+	}
+	return b.String()
+}
+
+// pad appends given str to the string buffer.
+func pad(b *strings.Builder, str string) {
+	if b.Len() == 0 {
+		return
+	}
+	b.WriteString(str)
+}
+
+func (e *Error) Zero() bool {
+	return e.Op == "" && e.Path == "" && e.Kind == 0 && e.Err == nil
+}
+
+// Path describes the path involved in the operation for ex. package
+// path in a "get" operation. Note that path could be a OS-specific or
+// OS-agnostic path depending on the operation being performed.
+type Path string
+
+// Op describes the operation being performed.
+type Op string
+
+// Kind describes the class of errors encountered.
+type Kind int
+
+const (
+	Other        Kind = iota // Unclassified. Will not be printed.
+	Exist                    // Item already exists.
+	Internal                 // Internal error.
+	InvalidParam             // Value is not valid.
```

Comment:
These error types need to be well-documented in order for them to be used consistently.

What's a param? Is it user-facing or does this capture programmer error?
Or is programmer error "Internal"?

To disambiguate, provide a concrete example of each.

Reply:
I haven't really used this field much yet in hydration errors, @mortent is experimenting with it in packaging errors. I haven't formed a mental model yet about it. I can document it better once we are happy with it's usage (somewhere around m2 milestone).